pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/generator/DirectiveGenerator.java

```diff
@@ -0,0 +1,191 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.pipeline.modeldefinition.generator;
+
+import hudson.Extension;
+import hudson.Functions;
+import hudson.model.Action;
+import hudson.model.Descriptor;
+import hudson.model.DescriptorByNameOwner;
+import hudson.model.Item;
+import hudson.model.Job;
+import hudson.model.RootAction;
+import jenkins.model.Jenkins;
+import jenkins.model.TransientActionFactory;
+import net.sf.json.JSONObject;
+import org.jenkinsci.plugins.structs.SymbolLookup;
+import org.jenkinsci.plugins.structs.describable.UninstantiatedDescribable;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowExecution;
+import org.jenkinsci.plugins.workflow.cps.Snippetizer;
+import org.jenkinsci.plugins.workflow.steps.StepDescriptor;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.DoNotUse;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.HttpResponse;
+import org.kohsuke.stapler.HttpResponses;
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.StaplerRequest;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@Extension
+public class DirectiveGenerator implements RootAction, DescriptorByNameOwner {
+
+    public static final String ACTION_URL = "directive-generator";
+
+    @Restricted(NoExternalUse.class)
+    public static final String GENERATE_URL = ACTION_URL + "/generateDirective";
+
+    @Override public String getUrlName() {
+        return ACTION_URL;
+    }
+
+    @Override public String getIconFileName() {
+        return null;
+    }
+
+    @Override public String getDisplayName() {
+        // Do not want to add to main Jenkins sidepanel.
+        return null;
+    }
+
+    @Override public Descriptor getDescriptorByName(String id) {
+        return Jenkins.getActiveInstance().getDescriptorByName(id);
+    }
+
+    @Nonnull
+    public List<DirectiveDescriptor> getDirectives() {
+        return DirectiveDescriptor.all();
+    }
+
+    @Restricted(DoNotUse.class) // for stapler
+    public @CheckForNull
+    Item getItem(StaplerRequest req) {
```

Comment:
If it's for stapler I don't think package level access works. IIRC it needs to be public

Reply:
That's a weird copypasta - the `public` is on the line above. Fixing.